File: leetcode/integer_to_roman.py
# Given an integer, convert it to a roman numeral.
# Input is guaranteed to be within the range from 1 to 3999.
# Time complexity: O(n)
# Space complexity: O(output)

class Solution:
    # @param {integer} num
    # @return {string}
    def intToRoman(self, num):
        M = ["", "M", "MM", "MMM"]
        C = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"]
        X = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"]
        I = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"]
        return M[num / 1000] + C[(num % 1000) / 100] + \
               X[(num % 100) / 10] + I[num % 10]

if __name__ == "__main__":
    solution = Solution()
    print solution.intToRoman(3999)

File: leetcode/integer_to_roman.js
/* Given an integer, convert it to a roman numeral.
 * Input is guaranteed to be within the range from 1 to 3999.
 * Time complexity: O(n)
 * Space complexity: O(output)
 */

/**
 * @param {number} num
 * @return {string}
 */
var intToRoman = function(num) {
    var M = ["", "M", "MM", "MMM"];
    var C = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"];
    var X = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"];
    var I = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"];
    var iM = Math.floor(num / 1000);
    var iC = Math.floor((num % 1000) / 100);
    var iX = Math.floor((num % 100) / 10);
    var iI = Math.floor(num % 10);
    return M[iM].concat(C[iC], X[iX], I[iI]);
};
